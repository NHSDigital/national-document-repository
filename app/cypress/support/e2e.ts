/// <reference types="cypress" />

import {
    AttributeValue,
    BatchWriteItemCommandOutput,
    DeleteItemCommandOutput,
    PutItemCommandOutput,
} from '@aws-sdk/client-dynamodb';

import { DeleteObjectCommandOutput, PutObjectCommandOutput  } from '@aws-sdk/client-s3';
import { Roles, roleIds, roleList } from './roles';
import { routes } from './routes';
import { defaultFeatureFlags, FeatureFlags } from './feature_flags';
import './aws.commands';
import 'cypress-real-events';

const registerCypressGrep = require('@cypress/grep');
registerCypressGrep();

Cypress.Commands.add('getByTestId', (selector, ...args) => {
    return cy.get(`[data-testid=${selector}]`, ...args);
});

Cypress.Commands.add('login', (role, featureFlags) => {
    if (roleIds.includes(role)) {
        const roleName = roleList.find((roleName) => Roles[roleName] === role);
        // Login for regression tests
        const authCallback = '/auth-callback';
        const fixturePath = 'requests/auth/GET_TokenRequest_' + roleName + '.json';

        cy.intercept('GET', '/Auth/TokenRequest*', {
            statusCode: 200,
            fixture: fixturePath,
        }).as('auth');

        if (featureFlags) {
            cy.intercept('GET', '/FeatureFlags*', {
                statusCode: 200,
                body: featureFlags,
            }).as('featureFlags');
        } else {
            cy.intercept('GET', '/FeatureFlags*', {
                statusCode: 200,
                body: defaultFeatureFlags,
            }).as('featureFlags');
        }

        cy.visit(authCallback);
        cy.wait('@auth');
        cy.wait('@featureFlags');
    } else {
        throw new Error("Invalid role for login. Only 'gp' or 'pcse' are allowed.");
    }
});

Cypress.Commands.add('smokeLogin', (role) => {
    // Login for smoke tests
    const baseUrl = Cypress.config('baseUrl');
    const key = Cypress.env('KEY');
    const odsCode = Cypress.env('ODSCODE');

    const startUrl = '/';
    cy.visit(startUrl);
    cy.url().should('eq', baseUrl + startUrl);
    cy.get('.nhsuk-header__transactional-service-name--link').should(
        'have.text',
        'Access and store digital patient documents',
    );
    cy.get('.nhsuk-header__navigation').should('not.exist');
    cy.getByTestId('start-btn').should('exist');
    cy.getByTestId('start-btn').click();
    cy.get('#key').should('exist');
    cy.get('#key').type(key);
    cy.get('#odsCode').should('exist');
    cy.get('#odsCode').type(odsCode);
    cy.get('#repositoryRole').should('exist');
    cy.get('#repositoryRole').select(role);
    cy.get('#submit-login-details').should('exist');
    cy.get('#submit-login-details').click();
});

Cypress.Commands.add('navigateToHomePage', () => {
    const baseUrl = Cypress.config('baseUrl');

    cy.getByTestId('home-btn').click();
    cy.url().should('eq', baseUrl + routes.home);
});

Cypress.Commands.add('navigateToPatientSearchPage', () => {
    const baseUrl = Cypress.config('baseUrl');

    cy.navigateToHomePage();
    cy.getByTestId('search-patient-btn').should('exist');
    cy.getByTestId('search-patient-btn').click();

    cy.url().should('eq', baseUrl + routes.patientSearch);
});

Cypress.Commands.add('navigateToDownloadReportPage', () => {
    const baseUrl = Cypress.config('baseUrl');

    cy.navigateToHomePage();
    cy.getByTestId('download-report-btn').should('exist');
    cy.getByTestId('download-report-btn').click();

    cy.url().should('eq', baseUrl + `${routes.createReport}?reportType=0`);
});

declare global {
    namespace Cypress {
        interface Chainable {
            /**
             * Get DOM element by data-testid attribute.
             *
             * @param {string} value - The value of the data-testid attribute of the target DOM element.
             * @return {HTMLElement} - Target DOM element.
             */
            getByTestId(value: string): Chainable<JQuery<HTMLElement>>;
            /**
             * Mock user login by intercepting the {baseUrl}/auth-callback request
             * @param {Roles} role - The user role to login with. Must be an enum of Roles
             * @param featureFlags - Feature flags values to override the defaults
             */
            login(role: Roles, featureFlags?: FeatureFlags): Chainable<void>;

            /**
             * Real user login via CIS2 and redirect back to {baseUrl}/auth-callback.
             * @param {Roles} role - The user role to login with. Must be an enum of Roles
             */
            smokeLogin(role: Roles): Chainable<void>;
            /**
             * Add file to s3 bucket
             * @param {string} bucketName - Name of the target S3 bucket
             * @param {string} fileName - Filepath of the file to upload
             * @param {string} fileContent - Content of the file to upload
             * @return {Promise<SendData>} - S3 response for s3.upload
             */
            addPdfFileToS3(
                bucketName: string,
                fileName: string,
                filePath: string,
            ): Chainable<PutObjectCommandOutput>;
            /**
             * Add dynamoDB entry
             * @param {string} tableName - Name of the target dynamoDB table
             * @param {{ [key: string]: any; }} item - dynamoDB item to upload
             * @return {Promise<PutItemOutput>} - Dynamo response for dynamoDB.putItem
             */
            addItemToDynamoDb(
                tableName: string,
                item: Record<string, AttributeValue>,
            ): Chainable<PutItemCommandOutput>;
            /**
             * Delete file from S3 bucket
             * @param {string} bucketName - Name of the target S3 bucket
             * @param {string} fileName - Filepath of the file to delete
             * @return {Promise<DeleteObjectOutput>} - S3 response for s3.deleteObject
             */
            deleteFileFromS3(
                bucketName: string,
                fileName: string,
            ): Chainable<DeleteObjectCommandOutput>;
            /**
             * Delete item from DynamoDB table
             * @param {string} tableName - Name of the target DynamoDB table
             * @param {string} itemId - ID of the item to delete
             * @return {Promise<DeleteItemOutput>} - Dynamo response for dynamoDB.deleteItem
             */
            deleteItemFromDynamoDb(
                tableName: string,
                itemId: string,
            ): Chainable<DeleteItemCommandOutput>;
            /**
             * Delete items with a specific secondary key value from DynamoDB table
             * @param {string} tableName - Name of the target DynamoDB table
             * @param {string} index - Index for the secondary key
             * @param {string} attribute - Name of the attribute you are matching on
             * @param {string} value - Specific value you are matching on to delete
             * @return {Promise<BatchWriteItemOutput>} - Dynamo response for dynamoDB.BatchWriteItem
             */
            deleteItemsBySecondaryKeyFromDynamoDb(
                tableName: string,
                index: string,
                attribute: string,
                value: string,
            ): Chainable<void>;

            navigateToHomePage(): Chainable<void>;
            navigateToPatientSearchPage(): Chainable<void>;
            navigateToDownloadReportPage(): Chainable<void>;
        }
    }
}
