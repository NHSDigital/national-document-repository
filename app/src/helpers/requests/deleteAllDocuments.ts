import axios from 'axios';
import { AuthHeaders } from '../../types/blocks/authHeaders';
import { DOCUMENT_TYPE } from '../../types/pages/UploadDocumentsPage/types';

type Args = {
    docType: DOCUMENT_TYPE;
    nhsNumber: string;
    baseUrl: string;
    baseHeaders: AuthHeaders;
};

type DeleteResponse = {
    data: string;
    status: number;
};
const deleteAllDocuments = async ({ docType, nhsNumber, baseUrl, baseHeaders }: Args) => {
    const gatewayUrl = baseUrl + '/DocumentDelete';

    try {
        const response: DeleteResponse = await axios.delete(gatewayUrl, {
            headers: {
                ...baseHeaders,
            },
            params: {
                patientId: nhsNumber,
                docType,
            },
        });
        return response;
    } catch (e) {
        throw e;
    }
};

export default deleteAllDocuments;
