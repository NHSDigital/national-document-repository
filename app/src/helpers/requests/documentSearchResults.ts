import { endpoints } from '../../types/generic/endpoints';
import { SearchResult } from '../../types/generic/searchResult';

import axios from 'axios';

type Args = {
    nhsNumber: string;
    baseUrl: string;
};

type GetDocumentSearchResultsResponse = {
    data: Array<SearchResult>;
};

const getDocumentSearchResults = async ({ nhsNumber, baseUrl }: Args) => {
    const gatewayUrl = baseUrl + endpoints.DOCUMENT_SEARCH;

    const { data }: GetDocumentSearchResultsResponse = await axios.get(gatewayUrl, {
        headers: {
            'Content-Type': 'application/json',
        },
        params: {
            patientId: nhsNumber,
        },
    });
    return data;
};

export default getDocumentSearchResults;
