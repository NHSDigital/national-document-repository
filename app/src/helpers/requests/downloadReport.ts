import { AuthHeaders } from '../../types/blocks/authHeaders';
import axios, { AxiosError } from 'axios';
import { ReportData } from '../../types/generic/reports';

type Args = {
    report: ReportData;
    fileType: string;
    baseUrl: string;
    baseHeaders: AuthHeaders;
};

type DownloadReportResponseData = {
    data: { url: string };
};

const downloadReport = async ({ report, fileType, baseUrl, baseHeaders }: Args): Promise<void> => {
    const gatewayUrl = baseUrl + report.endpoint;

    try {
        const { data }: DownloadReportResponseData = await axios.get(gatewayUrl, {
            headers: {
                ...baseHeaders,
            },
            params: {
                outputFileFormat: fileType,
            },
        });

        const link = document.createElement('a');
        link.href = data.url;
        link.setAttribute('download', '');

        link.click();
    } catch (e) {
        const error = e as AxiosError;
        throw error;
    }
};

export default downloadReport;
