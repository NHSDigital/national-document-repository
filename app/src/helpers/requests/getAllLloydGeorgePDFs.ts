import { AxiosError } from 'axios';
import axios from 'axios';
import { endpoints } from '../../types/generic/endpoints';
import { AuthHeaders } from '../../types/blocks/authHeaders';

type Args = {
    baseUrl: string;
    baseHeaders: AuthHeaders;
    nhsNumber: string;
};

const getAllLloydGeorgePDFs = async ({ baseUrl, baseHeaders, nhsNumber }: Args) => {
    const gatewayUrl = baseUrl + endpoints.LG_STITCH;
    try {
        const response = await axios.get(gatewayUrl, {
            headers: {
                ...baseHeaders,
            },
            params: {
                patientId: nhsNumber,
            },
        });
        return response;
    } catch (e) {
        const error = e as AxiosError;
        throw error;
    }
};

export default getAllLloydGeorgePDFs;
