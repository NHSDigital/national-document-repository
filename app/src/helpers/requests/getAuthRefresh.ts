import { AxiosInstance } from 'axios';
import { endpoints } from '../../types/generic/endpoints';
import { UserAuth } from '../../types/blocks/userAuth';

export type AuthTokenArgs = {
    refreshToken: string;
    axios: AxiosInstance;
};

const getAuthRefresh = async ({ refreshToken, axios }: AuthTokenArgs) => {
    try {
        const { data } = await axios.get(endpoints.REFRESH_AUTH, {
            params: { refreshToken },
        });
        const userAuth: UserAuth = data;
        return userAuth;
    } catch (e) {
        throw e;
    }
};

export default getAuthRefresh;
