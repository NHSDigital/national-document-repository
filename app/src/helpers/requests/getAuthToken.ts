import axios from 'axios';
import { endpoints } from '../../types/generic/endpoints';
import { UserAuth } from '../../types/blocks/userAuth';

export type AuthTokenArgs = {
    baseUrl: string;
    code: string;
    state: string;
};

const getAuthToken = async ({ baseUrl, code, state }: AuthTokenArgs) => {
    try {
        const tokenResponse: UserAuth = await axios.get(`${baseUrl}${endpoints.AUTH}`, {
            params: { code, state },
        });
        return tokenResponse;
    } catch (e) {
        throw e;
    }
};

export default getAuthToken;
