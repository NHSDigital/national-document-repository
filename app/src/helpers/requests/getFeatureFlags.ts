import { AuthHeaders } from '../../types/blocks/authHeaders';
import { endpoints } from '../../types/generic/endpoints';

import axios from 'axios';
import { FeatureFlags } from '../../types/generic/featureFlags';

type Args = {
    baseUrl: string;
    baseHeaders: AuthHeaders;
};

type GetFeatureFlagsResponse = {
    data: FeatureFlags;
};

export const defaultFeatureFlags = {
    testFeature1: false,
    testFeature2: false,
    testFeature3: false,
};

const getFeatureFlags = async ({ baseUrl, baseHeaders }: Args) => {
    const gatewayUrl = baseUrl + endpoints.FEATURE_FLAGS;
    try {
        const { data }: GetFeatureFlagsResponse = await axios.get(gatewayUrl, {
            headers: {
                ...baseHeaders,
            },
        });
        return data;
    } catch (e) {
        return defaultFeatureFlags;
    }
};

export default getFeatureFlags;
