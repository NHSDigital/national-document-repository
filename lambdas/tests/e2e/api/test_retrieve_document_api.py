import io
import os
import uuid

import requests
from tests.e2e.conftest import LLOYD_GEORGE_SNOMED, API_KEY, API_ENDPOINT
from tests.e2e.helpers.lloyd_george_data_helper import LloydGeorgeDataHelper

data_helper = LloydGeorgeDataHelper()


def test_small_file(test_data, snapshot):
    lloyd_george_record = {}
    test_data.append(lloyd_george_record)

    lloyd_george_record["id"] = str(uuid.uuid4())
    lloyd_george_record["nhs_number"] = "9449305943"
    lloyd_george_record["data"] = io.BytesIO(b"Sample PDF Content")

    data_helper.create_metadata(lloyd_george_record)
    data_helper.create_resource(lloyd_george_record)

    url = f"https://{API_ENDPOINT}/FhirDocumentReference/{LLOYD_GEORGE_SNOMED}~{lloyd_george_record['id']}"
    headers = {"Authorization": "Bearer 123", "X-Api-Key": API_KEY}
    response = requests.request("GET", url, headers=headers)
    json = response.json()

    del json["date"]
    del json["id"]

    assert json == snapshot


def test_large_file(test_data, snapshot):
    lloyd_george_record = {}
    test_data.append(lloyd_george_record)

    s3_bucket_name = os.environ.get("NDR_S3_BUCKET") or ""

    lloyd_george_record["id"] = str(uuid.uuid4())
    lloyd_george_record["nhs_number"] = "9449305943"
    lloyd_george_record["data"] = io.BytesIO(b"A" * (10 * 1024 * 1024))
    lloyd_george_record["size"] = 10 * 1024 * 1024 * 1024

    data_helper.create_metadata(lloyd_george_record)
    data_helper.create_resource(lloyd_george_record)

    url = f"https://{API_ENDPOINT}/FhirDocumentReference/{LLOYD_GEORGE_SNOMED}~{lloyd_george_record['id']}"
    headers = {"Authorization": "Bearer 123", "X-Api-Key": API_KEY}
    response = requests.request("GET", url, headers=headers)
    json = response.json()

    expected_presign_uri = (
        f"https://{s3_bucket_name}.s3.eu-west-2.amazonaws.com/{lloyd_george_record['nhs_number']}/{lloyd_george_record['id']}"
    )
    assert expected_presign_uri in json["content"][0]["attachment"]["url"]

    del json["date"]
    del json["id"]
    del json["content"][0]["attachment"]["url"]

    assert json == snapshot


def test_no_file_found(snapshot):
    lloyd_george_record = {}
    lloyd_george_record["id"] = str(uuid.uuid4())

    url = f"https://{API_ENDPOINT}/FhirDocumentReference/{LLOYD_GEORGE_SNOMED}~{lloyd_george_record['id']}"
    headers = {"Authorization": "Bearer 123", "X-Api-Key": API_KEY}
    response = requests.request("GET", url, headers=headers)
    json = response.json()

    assert json == snapshot


def test_preliminary_file(test_data, snapshot):
    lloyd_george_record = {}
    test_data.append(lloyd_george_record)

    lloyd_george_record["id"] = str(uuid.uuid4())
    lloyd_george_record["nhs_number"] = "9449305943"
    lloyd_george_record["data"] = io.BytesIO(b"Sample PDF Content")
    lloyd_george_record["doc_status"] = "preliminary"

    data_helper.create_metadata(lloyd_george_record)
    data_helper.create_resource(lloyd_george_record)

    url = f"https://{API_ENDPOINT}/FhirDocumentReference/{LLOYD_GEORGE_SNOMED}~{lloyd_george_record['id']}"
    headers = {"Authorization": "Bearer 123", "X-Api-Key": API_KEY}
    response = requests.request("GET", url, headers=headers)
    json = response.json()

    del json["date"]
    del json["id"]

    assert json == snapshot
