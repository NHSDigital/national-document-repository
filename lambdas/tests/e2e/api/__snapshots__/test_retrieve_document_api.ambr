# serializer version: 1
# name: test_large_file
  dict({
    'author': list([
      dict({
        'identifier': dict({
          'system': 'https://fhir.nhs.uk/Id/ods-organization-code',
          'value': 'H81109',
        }),
      }),
    ]),
    'content': list([
      dict({
        'attachment': dict({
          'contentType': 'application/pdf',
          'creation': '2023-01-01',
          'language': 'en-GB',
          'size': 10737418240,
          'title': '1of1_Lloyd_George_Record_[Holly Lorna MAGAN]_[9449305943]_[29-05-2006].pdf',
        }),
      }),
    ]),
    'custodian': dict({
      'identifier': dict({
        'system': 'https://fhir.nhs.uk/Id/ods-organization-code',
        'value': 'H81109',
      }),
    }),
    'docStatus': 'final',
    'resourceType': 'DocumentReference',
    'status': 'current',
    'subject': dict({
      'identifier': dict({
        'system': 'https://fhir.nhs.uk/Id/nhs-number',
        'value': '9449305943',
      }),
    }),
    'type': dict({
      'coding': list([
        dict({
          'code': '16521000000101',
          'display': 'Lloyd George record folder',
          'system': 'http://snomed.info/sct',
        }),
      ]),
    }),
  })
# ---
# name: test_no_file_found
  dict({
    'issue': list([
      dict({
        'code': 'exception',
        'details': dict({
          'coding': list([
            dict({
              'code': 'RESOURCE_NOT_FOUND',
              'display': 'Resource not found',
              'system': 'https://fhir.hl7.org.uk/CodeSystem/UKCore-SpineErrorOrWarningCode',
            }),
          ]),
        }),
        'diagnostics': 'Document reference not found',
        'severity': 'error',
      }),
    ]),
    'resourceType': 'OperationOutcome',
  })
# ---
# name: test_preliminary_file
  dict({
    'author': list([
      dict({
        'identifier': dict({
          'system': 'https://fhir.nhs.uk/Id/ods-organization-code',
          'value': 'H81109',
        }),
      }),
    ]),
    'content': list([
      dict({
        'attachment': dict({
          'contentType': 'application/pdf',
          'creation': '2023-01-01',
          'language': 'en-GB',
          'title': '1of1_Lloyd_George_Record_[Holly Lorna MAGAN]_[9449305943]_[29-05-2006].pdf',
        }),
      }),
    ]),
    'custodian': dict({
      'identifier': dict({
        'system': 'https://fhir.nhs.uk/Id/ods-organization-code',
        'value': 'H81109',
      }),
    }),
    'docStatus': 'preliminary',
    'resourceType': 'DocumentReference',
    'status': 'current',
    'subject': dict({
      'identifier': dict({
        'system': 'https://fhir.nhs.uk/Id/nhs-number',
        'value': '9449305943',
      }),
    }),
    'type': dict({
      'coding': list([
        dict({
          'code': '16521000000101',
          'display': 'Lloyd George record folder',
          'system': 'http://snomed.info/sct',
        }),
      ]),
    }),
  })
# ---
# name: test_small_file
  dict({
    'author': list([
      dict({
        'identifier': dict({
          'system': 'https://fhir.nhs.uk/Id/ods-organization-code',
          'value': 'H81109',
        }),
      }),
    ]),
    'content': list([
      dict({
        'attachment': dict({
          'contentType': 'application/pdf',
          'creation': '2023-01-01',
          'data': 'U2FtcGxlIFBERiBDb250ZW50',
          'language': 'en-GB',
          'size': 12345,
          'title': '1of1_Lloyd_George_Record_[Holly Lorna MAGAN]_[9449305943]_[29-05-2006].pdf',
        }),
      }),
    ]),
    'custodian': dict({
      'identifier': dict({
        'system': 'https://fhir.nhs.uk/Id/ods-organization-code',
        'value': 'H81109',
      }),
    }),
    'docStatus': 'final',
    'resourceType': 'DocumentReference',
    'status': 'current',
    'subject': dict({
      'identifier': dict({
        'system': 'https://fhir.nhs.uk/Id/nhs-number',
        'value': '9449305943',
      }),
    }),
    'type': dict({
      'coding': list([
        dict({
          'code': '16521000000101',
          'display': 'Lloyd George record folder',
          'system': 'http://snomed.info/sct',
        }),
      ]),
    }),
  })
# ---
